modules/core/src/lib/widget-manager.ts

```diff
@@ -57,61 +63,119 @@ const ROOT_CONTAINER_ID = '__root';
 export class WidgetManager {
   deck: Deck;
   parentElement?: HTMLElement | null;
-  containers: {[id: string]: HTMLDivElement} = {};
-  widgets: Widget[] = [];
-  lastViewports: {[id: string]: Viewport} = {};
+
+  /** Widgets added via the imperative API */
+  private defaultWidgets: Widget[] = [];
+  /** Widgets received from the declarative API */
+  private widgets: Widget[] = [];
+  /** Resolved widgets from both imperative and declarative APIs */
+  private resolvedWidgets: Widget[] = [];
+  private containers: {[id: string]: HTMLDivElement} = {};
+  private lastViewports: {[id: string]: Viewport} = {};
 
   constructor({deck, parentElement}: {deck: Deck; parentElement?: HTMLElement | null}) {
     this.deck = deck;
     this.parentElement = parentElement;
   }
 
+  getWidgets(): Widget[] {
+    return this.resolvedWidgets;
+  }
+
+  /** Declarative API to configure widgets */
+  setProps(props: {widgets?: Widget[]}) {
+    if (props.widgets && !deepEqual(props.widgets, this.widgets, 1)) {
+      this._setWidgets(props.widgets);
+    }
+  }
+
   finalize() {
-    for (const widget of this.widgets) {
-      this.remove(widget);
+    for (const widget of this.getWidgets()) {
+      this._remove(widget);
     }
+    this.defaultWidgets.length = 0;
+    this.resolvedWidgets.length = 0;
     for (const id in this.containers) {
       this.containers[id].remove();
     }
   }
 
-  add(
-    widget: Widget,
-    opts: {
-      viewId?: string | null;
-      placement?: WidgetPlacement;
-    } = {}
-  ) {
-    if (this.widgets.includes(widget)) {
-      // widget already added
-      return;
+  /** Imperative API. Widgets added this way are not affected by the declarative prop. */
+  addDefault(widget: Widget) {
```

Comment:
I was about to suggest a different name, but I guess this was renamed to addDefault based on previous comments.